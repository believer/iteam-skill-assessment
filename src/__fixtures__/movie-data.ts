import { build, fake } from 'test-data-bot'
import { SearchMovie, MovieType, Movie } from '../types'

type PartialMovie = Pick<
  Movie,
  'Actors' | 'Poster' | 'Title' | 'Type' | 'Year' | 'imdbID'
>

export const searchMovieBuilder = build<SearchMovie>('Movie').fields({
  Poster: fake(f => f.internet.url()),
  Title: fake(f => f.lorem.words()),
  Type: MovieType.Movie,
  Year: fake(f =>
    f.date
      .past()
      .getFullYear()
      .toString()
  ),
  imdbID: fake(f => f.random.alphaNumeric()),
})

export const movieBuilder = build<PartialMovie>('Movie').fields({
  Actors: 'test, test2, test3',
  Poster: fake(f => f.internet.url()),
  Title: fake(f => f.lorem.words()),
  Type: MovieType.Movie,
  Year: fake(f =>
    f.date
      .past()
      .getFullYear()
      .toString()
  ),
  imdbID: fake(f => f.random.alphaNumeric()),
})
