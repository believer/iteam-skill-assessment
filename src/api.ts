import useFetch from 'react-fetch-hook'
import { MovieResponse, SearchResults } from './types'

const API_KEY = process.env.REACT_APP_OMDB_KEY

const createRequest = (params: string) =>
  `https://www.omdbapi.com/?apikey=${API_KEY}${params}`

export const useSearch = (query: string) => {
  const { data, isLoading, error } = useFetch<SearchResults>(
    createRequest(`&s=${query}`),
    {
      depends: [query],
    }
  )

  if (!data) {
    return { movies: [], isLoading, error }
  }

  return { movies: data.Search || [], isLoading, error }
}

export const useMovie = (id?: string) => {
  const { data, isLoading, error } = useFetch<MovieResponse>(
    createRequest(`&i=${id}`),
    {
      depends: [id],
    }
  )

  if (!data) {
    return { movie: null, isLoading, error }
  }

  if (data.Error) {
    return { movie: null, isLoading, error: data.Error }
  }

  return {
    movie: data,
    isLoading,
    error,
  }
}
