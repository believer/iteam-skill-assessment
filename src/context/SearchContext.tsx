import React from 'react'

export const SearchContext = React.createContext({
  query: '',
  setQuery: (_value: string) => {},
})

export const SearchProvider: React.FC = ({ children }) => {
  const [query, setQuery] = React.useState('')

  return (
    <SearchContext.Provider value={{ query, setQuery }}>
      {children}
    </SearchContext.Provider>
  )
}
