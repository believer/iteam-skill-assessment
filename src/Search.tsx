import styled from '@emotion/styled'
import { useDebounce } from '@iteam/hooks'
import { RouteComponentProps } from '@reach/router'
import * as React from 'react'
import { useSearch } from './api'
import iconSearch from './images/i_search.svg'
import MovieResult from './MovieResult'
import { Grid, GridColumn } from './Grid'
import { SearchContext } from './SearchContext'
import { LoadingBlock } from './LoadingBlock'

interface SearchInputProps {
  hasQuery: boolean
}

const SearchWrap = styled.div<SearchInputProps>`
  background-color: hsla(0, 0%, 90%, 0.8);
  display: flex;
  justify-content: center;
  left: 0;
  padding: 20px;
  position: fixed;
  right: 0;
  transform: translateY(${({ hasQuery }) => (hasQuery ? '0' : '-50%')});
  transition: 200ms ease-in-out;
  transition-property: transform, top, border-color;
  top: ${({ hasQuery }) => (hasQuery ? '0' : '50%')};
  z-index: 1;
`

const SearchInput = styled.input<SearchInputProps>`
  background-color: transparent;
  background-image: url(${iconSearch});
  background-repeat: no-repeat;
  background-size: 24px; 24px;
  background-position: 0 center;
  border: 0;
  border-bottom: 2px solid #009eb5;
  font-family: 'Fjalla One', sans-serif;
  font-size: 32px;
  padding: 10px;
  padding-left: 40px;
  width: 80vw;

  @media (min-width: 52em) {
    width: 400px;
  }
`

const Results = styled.div`
  display: grid;
  grid-gap: 20px;
  grid-template-columns: repeat(2, 1fr);
  margin-top: 140px;

  @media (min-width: 52em) {
    grid-gap: 40px;
    grid-template-columns: repeat(4, 1fr);
    margin-top: 200px;
  }
`

const Loading = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
`

const Search: React.FC<RouteComponentProps> = () => {
  const { query, setQuery } = React.useContext(SearchContext)
  const debouncedValue = useDebounce(query, 300)
  const { movies, isLoading } = useSearch(debouncedValue)

  const handleChange = e => setQuery(e.currentTarget.value)

  return (
    <Grid mb={[40, 100]}>
      <SearchWrap hasQuery={query.length > 0}>
        <SearchInput
          aria-label="Find a movie you love"
          hasQuery={query.length > 0}
          onChange={handleChange}
          placeholder="Find a movie you love"
          type="text"
          value={query}
        />
      </SearchWrap>

      <GridColumn>
        <Results>
          {isLoading &&
            Array.from(Array(12).keys()).map(i => (
              <Loading key={i}>
                <LoadingBlock height="315px" />
                <LoadingBlock height="19px" mt="15px" width="80%" />
              </Loading>
            ))}

          {movies.map(movie => (
            <MovieResult key={movie.id} movie={movie} />
          ))}
        </Results>
      </GridColumn>
    </Grid>
  )
}

export default Search
