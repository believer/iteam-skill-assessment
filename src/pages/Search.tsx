import styled from '@emotion/styled'
import { useDebounce } from '@iteam/hooks'
import { RouteComponentProps } from '@reach/router'
import * as React from 'react'
import { useSearch } from '../api'
import { LoadingBlock } from '../components/LoadingBlock'
import MovieResult from '../components/MovieResult'
import { SearchContext } from '../context/SearchContext'
import iconSearch from '../images/i_search.svg'

interface SearchInputProps {
  hasQuery: boolean
}

const SearchWrap = styled.div<SearchInputProps>`
  left: 0;
  right: 0;
  transform: translateY(${({ hasQuery }) => (hasQuery ? '0' : '-50%')});
  transition: 200ms ease-in-out;
  transition-property: transform, top, border-color;
  top: ${({ hasQuery }) => (hasQuery ? '0' : '50%')};
  z-index: 1;
`

const SearchInput = styled.input<SearchInputProps>`
  background-color: transparent;
  background-image: url(${iconSearch});
  background-repeat: no-repeat;
  background-size: 24px 24px;
  background-position: 0 center;
  font-size: 32px;
  padding: 10px;
  padding-left: 40px;
  width: 80vw;

  @media (min-width: 52em) {
    width: 400px;
  }
`

const Results = styled.div`
  display: grid;
  grid-gap: 20px;
  grid-template-columns: repeat(2, 1fr);
  margin-top: 140px;

  @media (min-width: 52em) {
    grid-gap: 40px;
    grid-template-columns: repeat(4, 1fr);
    margin-top: 200px;
  }
`
const Search: React.FC<RouteComponentProps> = () => {
  const { query, setQuery } = React.useContext(SearchContext)
  const debouncedValue = useDebounce(query, 300)
  const { movies, isLoading } = useSearch(debouncedValue)

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setQuery(e.currentTarget.value)

  return (
    <div className="grid grid-template">
      <SearchWrap
        className="bg-gray-100 fixed flex justify-center p-4"
        hasQuery={query.length > 0}
      >
        <SearchInput
          aria-label="Find a movie you love"
          className="font-header border-b-2 border-gray-400 focus:border-teal-400 outline-none"
          hasQuery={query.length > 0}
          onChange={handleChange}
          placeholder="Find a movie you love"
          type="text"
          value={query}
        />
      </SearchWrap>

      <div className="grid-template-center">
        <Results>
          {isLoading &&
            Array.from(Array(12).keys()).map(i => (
              <div className="flex items-center flex-col" key={i}>
                <LoadingBlock height="315px" />
                <LoadingBlock height="19px" mt="15px" width="80%" />
              </div>
            ))}

          {!!query &&
            !isLoading &&
            movies.map(movie => (
              <MovieResult key={movie.imdbID} movie={movie} />
            ))}
        </Results>

        {!!debouncedValue && !isLoading && movies.length === 0 && (
          <div className="text-center text-2xl text-gray-600">
            No search results, try a different movie!
          </div>
        )}
      </div>
    </div>
  )
}

export default Search
