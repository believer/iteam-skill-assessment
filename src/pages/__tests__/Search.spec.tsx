import { act, fireEvent, render, wait } from '@testing-library/react'
import React from 'react'
import { useSearch } from '../../api'
import { SearchProvider } from '../../context/SearchContext'
import Search from '../Search'
import { searchMovieBuilder } from '../../__fixtures__/movie-data'

jest.mock('../../api')

test('renders no search results', async () => {
  jest.useFakeTimers()
  ;(useSearch as jest.Mock).mockReturnValue({
    movies: [],
    isLoading: false,
    error: null,
  })

  const { getByText, getByPlaceholderText } = render(
    <SearchProvider>
      <Search />
    </SearchProvider>
  )

  // Tests search
  fireEvent.change(getByPlaceholderText(/find a movie you love/i), {
    target: { value: 'batman' },
  })

  await wait()

  act(() => {
    jest.runAllTimers()
  })

  expect(getByText(/no search results/i)).toBeInTheDocument()
})

test('renders loading state', async () => {
  jest.useFakeTimers()
  ;(useSearch as jest.Mock).mockReturnValue({
    movies: [],
    isLoading: true,
    error: null,
  })

  const { container, getByPlaceholderText } = render(
    <SearchProvider>
      <Search />
    </SearchProvider>
  )

  // Tests search
  fireEvent.change(getByPlaceholderText(/find a movie you love/i), {
    target: { value: 'batman' },
  })

  await wait()

  act(() => {
    jest.runAllTimers()
  })

  expect(container).toMatchSnapshot()
})

test('renders and displays search results', async () => {
  const movie = searchMovieBuilder()

  jest.useFakeTimers()
  ;(useSearch as jest.Mock).mockReturnValue({
    movies: [movie],
    isLoading: false,
    error: null,
  })

  const { getByText, getByPlaceholderText, getByAltText } = render(
    <SearchProvider>
      <Search />
    </SearchProvider>
  )

  // Tests search
  fireEvent.change(getByPlaceholderText(/find a movie you love/i), {
    target: { value: 'batman' },
  })

  await wait()

  act(() => {
    jest.runAllTimers()
  })

  expect(useSearch).toHaveBeenCalledWith('batman')

  expect(getByText(movie.Title)).toBeInTheDocument()
  expect(getByAltText(movie.Title)).toBeInTheDocument()
})
