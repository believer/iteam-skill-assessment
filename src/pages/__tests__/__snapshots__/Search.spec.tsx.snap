// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders loading state 1`] = `
<div>
  <div
    class="grid grid-template"
  >
    <div
      class="bg-gray-100 fixed flex justify-center p-4 z-10 transition-fast transition-ease-in-out css-1x9gb4u"
    >
      <input
        aria-label="Find a movie you love"
        class="font-header border-b-2 border-gray-400 focus:border-teal-400 outline-none p-3 pl-12 text-3xl bg-transparent bg-no-repeat transition-fast transition-ease-in-out css-ctpkuz"
        placeholder="Find a movie you love"
        type="text"
        value="batman"
      />
    </div>
    <div
      class="grid-template-center"
    >
      <div
        class="grid mt-32 md:mt-48 css-nfkm8s"
      >
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
        <div
          class="flex items-center flex-col"
        >
          <div
            class="rounded bg-gray-200 h-80 w-full"
          />
          <div
            class="rounded bg-gray-200 h-5 mt-4 w-4/5"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
